### PR TITLE
[10.0][ADD] sale_blanket_order

### LINKED ISSUE #592: [10.0] New feature: blanket sales order management
One of my customers needs to manage blanket sales orders. I cannot find any thing fitting my needs neither in core neither in OCA. So I am planning to implement a new simple module.

My idea is to create a new object "Blanket Order" with these attributes:

- Customer
- Pricelist
- Payment terms
- Validity date
- State (Draft, Opened, Expired)
- List of products with proposed quantity and price

The workflow is:

1. Blanket Order Creation 
The user encodes the blanket order and once he finds an agreement with the customer, he confirms it

2. Consumption of quantities
Once it is confirmed, a wizard is available to generate sale orders. The user can select the product in the list and set the ordered quantity, which can not exceed the blanket order quantity.

3. End of validity

If all proposed quantities has been consumed, no new sale order can be created.
If the validity date is past, the blanket sale orders jump to "Expired" state whatever all quantities are consummed or not

At all moment, the user can know for all products:

- the original quantity
- the ordered quantity
- the delivered quantity
- the invoiced quantity

and can access to linked sales orders.

Does anyone has on opinion on my suggestion ?

### PR DESCRIPTION
**Pull request by andreparames**
This module adds blanket orders, as described in the corresponding issue.

This PR depends on another for the migration of `web_action_conditionable`: https://github.com/OCA/web/pull/845

Fixes #592

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please avoid plurals on module name.

File: sale_blanket_orders/__manifest__.py
Comment (Cedric-Pigeon) [inline sale_blanket_orders/__manifest__.py]: Version should be 10

File: sale_blanket_orders/__manifest__.py
Comment (Cedric-Pigeon) [inline sale_blanket_orders/__manifest__.py]: Acsone SA/NV

File: sale_blanket_orders/models/blanket_orders.py
Comment (Cedric-Pigeon) [inline sale_blanket_orders/models/blanket_orders.py]: Perhaps you should add the delivered quantities also

File: sale_blanket_orders/views/blanket_orders.xml
Comment (Cedric-Pigeon) [inline sale_blanket_orders/views/blanket_orders.xml]: should be set required in draft state through an attrs attribute

File: sale_blanket_orders/views/blanket_orders.xml
Comment (Cedric-Pigeon) [inline sale_blanket_orders/views/blanket_orders.xml]: should be set required in draft state

**Review by Cedric-Pigeon — CHANGES_REQUESTED** [review]

File: sale_blanket_orders/views/blanket_orders.xml
Comment (andreparames) [reply sale_blanket_orders/views/blanket_orders.xml]: No problem. I had added it initially, but then thought it could be useful to create the order before the validity date is agreed, so instead I added a check when the user confirms it.
I'll add them back.

File: sale_blanket_order/data/ir_cron.xml
Comment (atchuthan) [inline sale_blanket_order/data/ir_cron.xml]: should be noupdate="1"

**Review by atchuthan — APPROVED** [review]
Code review :+1:

**Comment by andreparames** [comment]
Thanks @pedrobaeza @Cedric-Pigeon @atchuthan for the review. The checks are failing because the module depends on `web_action_conditionable`, which I've migrated and is also waiting to be merged: https://github.com/OCA/web/pull/845

File: sale_blanket_order/models/blanket_orders.py
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/blanket_orders.py]: you should use `line.` instead of `self`

**Review by Cedric-Pigeon — APPROVED** [review]
Code review and functionnal tests

**Review by misern2 — CHANGES_REQUESTED** [review]
Thanks for the contribution.
Just a minor change:
 if you duplicate the blanket order it keeps the same sequence and state.

**Comment by andreparames** [comment]
Thanks @misern2, I've pushed a commit that fixes that problem, and also lets the lines be copied when duplicating an order.

**Review by misern2 — APPROVED** [review]
:+1: LGTM . Thanks!.

**Comment by andhit-r** [comment]
@andreparames thanks for contribution. Few things to consider:

Sequence
-------------

Since sequence will be assigned when blanket order confirmed, it will leave draft BO with **draft** identifier. IMO, it will give user a problem to identify draft BO.

**Propose Solution**

1. If you still want to assign sequence when BO confirmed, make display_name *{ID} like in account.move. Or,
2. Assign sequence when BO saved. With default name value == "/".

State
-------

It always useful to have approval step after confirm

**Propose Solution**

Draft -> Waiting for Approval -> Opened

If you agree my points above, i will create PR for those propose solutions.

Many thanks.

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: That's default

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: You can fetch them through one mapped('line_ids.sale_order_lines_ids.order_id')

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: Return just the result of mapped.

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py:81]: Is that correct ?

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: Deprecated, use UserError

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: Maybe 'open' is more correct

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: Readonly=True is recommended for related fields

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/blanket_orders.py]: Deprecated

File: sale_blanket_order/models/sale_orders.py
Comment (rousseldenis) [inline sale_blanket_order/models/sale_orders.py]: readonly=True

File: sale_blanket_order/wizard/create_sale_orders.py
Comment (rousseldenis) [inline sale_blanket_order/wizard/create_sale_orders.py]: Deprecated

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Some comments.

One remark : why don't you use decimal precision on quantities ?

File: sale_blanket_order/models/blanket_orders.py
Comment (andreparames) [reply sale_blanket_order/models/blanket_orders.py]: "Explicit is better than implicit" :smile:

**Comment by andreparames** [comment]
Thanks for the review @rousseldenis!

**Comment by andreparames** [comment]
@andhit-r they seem like good suggestions. I'd just add an option to enable the approval step.

**Review by rousseldenis — APPROVED** [review]

**Comment by heathfx** [comment]
I stumbled across this module while trolling the web and hope I'm not intruding by commenting here. This module could be very useful at my company and when I was testing it, noticed that the price compute method is likely missing an onchange decorator for when qty is updated.

This is important when pricelists are in use and different price tiers are based on qty.

**Comment by andreparames** [comment]
@heathfx good point, I've made the onchange recalculate the price when the initial quantity changes.

**Comment by andreparames** [comment]
I've pushed an update that adds a configuration option to allow or disallow adding new lines to SOs created from BOs (by default it allows).

File: sale_blanket_order/views/blanket_orders.xml
Comment (AdriaGForgeFlow) [inline sale_blanket_order/views/blanket_orders.xml:44]: I think it should be set to required in draft state to make it clear for the user. In the _validate (blanket_orders.py) function you specify that Partner field is mandatory but that field does not appear in purple color in the web interface and that's a little bit confusing IMO.

**Comment by bjeficent** [comment]
Just a question:
If you Create a blanket order > Create sale order and instead of confirm it you cancel it and then remove it, the ordered quantity of that sale order and the "updated" remaining quantity is shown. Is this how is it supposed to work?

File: sale_blanket_order/models/blanket_orders.py
Comment (JordiBForgeFlow) [inline sale_blanket_order/models/blanket_orders.py:304]: I think that you need to add api.depends('sale_order_lines_ids', 'sale_order_lines_ids.ordered_qty',  'sale_order_lines_ids.invoiced_qty', 'sale_order_lines_ids.delivered_qty'. Otherwise there will always be discrepancies.

**Comment by JordiBForgeFlow** [comment]
@bjeficent You are right. See my comment. Perhaps you can propose the fix.

File: sale_blanket_order/models/blanket_orders.py
Comment (bjeficent) [reply sale_blanket_order/models/blanket_orders.py:304]: You are right, those lines need to be added but without the sale_order_lines_ids in ordered_qty, invoiced_qty... I mean:
api.depends('sale_order_lines_ids', 'ordered_qty', 'invoiced_qty', 'delivered_qty', 'remaining_qty')

File: sale_blanket_order/README.rst
Comment (nikul-serpentcs) [inline sale_blanket_order/README.rst:22]: @andreparames `target` 10.0 :smiley:

File: sale_blanket_order/__manifest__.py
Comment (nikul-serpentcs) [inline sale_blanket_order/__manifest__.py:16]: @andreparames Sequence should be [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

File: sale_blanket_order/models/blanket_orders.py
Comment (bjeficent) [inline sale_blanket_order/models/blanket_orders.py:199]: I think the store=True parameter should be removed from ordered_qty, invoiced_qty, remaining_qty and delivered_qty to fix that issue.

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

**Comment by rgl1234** [comment]
At the moment we see the following missing features for this functionality:

1.) possibility to have an analytic-account on BO if needed

2.) In our case, the products and quantities can change during the lifetime of a blanket order. So it would be good if the blanket-order could be modified after approving it and after having started the deliveries on a BO (add/remove products, change quantities on products, change validity date...)

does this maybe make sence too for someone else?

René

**Comment by rgl1234** [comment]
At the moment we see the following missing features for this functionality:

1.) possibility to have an analytic-account on BO if needed

2.) In our case, the products and quantities can change during the lifetime of a blanket order. So it would be good if the blanket-order could be modified after approving it and after having started the deliveries on a BO (add/remove products, change quantities on products, change validity date...)

does this maybe make sence too for someone else?

René

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py:99]: If the order's validity date is today, then the order is considered expired? Is the equals needed?

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py]: Is it perhaps possible to just do `self.partner_id.property_product_pricelist.id` since this (despite it's name) is a `Many2one`? Same for the `property_payment_term_id`

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py]: From what I understand, here you are getting the default values of the fields `name` and `confirmed` and injecting them on the newly created record.

Is this override necessary though? Because `create()` calls the `_add_missing_default_values()` which in turn calls `default_get` and adds the values later. So removing this would make sense. Or am I missing something here?

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py:131]: Maybe this, with it's assertions, could be replaced by `@api.constrains('confirmed')` and conditionals?

So it does not explicitly requires us to call it and it would be called when necessary.
Also, there are certain flags passed to the interpreter which can remove the assertions

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py:161]: The `action['domain']` already has a domain which is removed when this assignment takes place. Perhaps it would be better to append it?

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py]: If you want to see whether your recordset has any records in it, you can avoid the `len()` altogether since `BaseModel.__nonzero__` is implemented. Therefore `if sale_orders` will suffice.

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py:313]: Same here, could this be perhaps replaced by `api.constrains()` and conditions instead of assert?

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py:207]: `pricelist_id` seems to not be used anywhere here.

File: sale_blanket_order/models/blanket_orders.py
Comment (daramousk) [inline sale_blanket_order/models/blanket_orders.py]: Perhaps `price_and_rule` should be renamed to `rule_id`?

**Review by daramousk — COMMENTED** [review]
Hello and thank you for this, looks great.
I have left some comments mostly about the assertions.

Let me know what you think.

**Comment by rafaelbn** [comment]
Hello @andreparames , please could you check bots are green :green_heart: ? I see travis red :red_circle:

**Comment by rousseldenis** [comment]
Hi @rafaelbn, @andreparames does not wor k anymore on Odoo. I will try to finish this one.

**Comment by rafaelbn** [comment]
Hi @rousseldenis , if you cannot work on this we can close this PR

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py:313]: I'll modify asserts but this cannot be constrains because it's called on order validation (action)

Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py:313]: And finally I think I will let asserts because I like the way it was done (catching AssertionError and raise an Odoo Exception)

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py:131]: See here under message

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py]: A simple copy=False will be sufficent

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py:199]: @bjeficent Why removing store=True ?

File: sale_blanket_order/models/blanket_orders.py
Comment (rousseldenis) [reply sale_blanket_order/models/blanket_orders.py:207]: Yes but can be overriden

File: sale_blanket_order/views/blanket_orders.xml
Comment (rousseldenis) [reply sale_blanket_order/views/blanket_orders.xml:44]: @Cedric-Pigeon What do you think of this ?

**Comment by rousseldenis** [comment]
@rafaelbn @daramousk Remarks that seems relevant to me has been attended. Waiting travis but ok for me.

File: sale_blanket_order/views/blanket_orders.xml
Comment (Cedric-Pigeon) [reply sale_blanket_order/views/blanket_orders.xml:44]: it makes sense IMO

**Comment by Cedric-Pigeon** [comment]
@rafaelbn This one is ready to be merged.

**Pull request merged**
